include/PluginPinConnector.h

```diff
@@ -0,0 +1,158 @@
+/*
+ * PluginPinConnector.h - Specifies how to route audio channels
+ *                        in and out of a plugin.
+ *
+ * Copyright (c) 2024 Dalton Messmer <messmer.dalton/at/gmail.com>
+ *
+ * This file is part of LMMS - https://lmms.io
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this program (see COPYING); if not, write to the
+ * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301 USA.
+ *
+ */
+
+#ifndef LMMS_PLUGIN_PIN_CONNECTOR_H
+#define LMMS_PLUGIN_PIN_CONNECTOR_H
+
+#include <vector>
+
+#include "AutomatableModel.h"
+#include "lmms_export.h"
+#include "SampleFrame.h"
+#include "SerializingObject.h"
+
+class QWidget;
+
+namespace lmms
+{
+
+namespace gui
+{
+
+class PluginPinConnectorView;
+
+} // namespace gui
+
+//! Configure channel routing for a plugin's mono/stereo in/out ports
+class LMMS_EXPORT PluginPinConnector
+	: public Model
+	, public SerializingObject
+{
+	Q_OBJECT
+
+public:
+	//! [LMMS track channel][plugin channel]
+	using PinMap = std::vector<std::vector<BoolModel*>>;
+
+	struct Matrix
+	{
+		explicit Matrix(bool isIn) : in{isIn} {}
+
+		auto channelName(int channel) const -> QString;
+
+		auto enabled(std::uint8_t trackChannel, unsigned pluginChannel) const -> bool
+		{
+			return pins[trackChannel][pluginChannel]->value();
+		}
+
+		PinMap pins;
+		int channelCount = 0;
+		std::vector<QString> channelNames; //!< optional
+		bool in; //!< true: LMMS-to-plugin; false: plugin-to-LMMS
+
+		// TODO: Channel groupings, port configurations, ...
+	};
+
+	PluginPinConnector(Model* parent = nullptr);
+	PluginPinConnector(int pluginInCount, int pluginOutCount, Model* parent = nullptr);
+
+	/**
+	 * Getters
+	 */
+	auto in() const -> const Matrix& { return m_in; };
```

Comment:
I just prefer using trailing return types. I think it's more readable - especially for long return types - and it has an improvement in the name lookup rules for nested types.

Reply:
There might be cases where it is necessary or improves something, but in most cases, it's equally worth, just more characters. Also, whenever I want to know the return type in this case, I will need to read through the whole function parameters until I reach the `->`.

There is probably no objective correct way, though it might be good to be consistent with other code. It might also be a good question to ask on Discord?

Reply:
It was discussed a while back on Discord. I don't think we officially decided on anything, but Dom and saker both like using it and so do I. When modifying a class that declares methods the typical way, I'll usually follow the convention that's already in use, but for new classes I like using the trailing return syntax.